### ISSUE TITLE
Inventory item name uniqueness not enforced with same parent device

### ISSUE THREAD

**Issue by aaron-iles** [netbox-community/netbox#13741]
### NetBox version

v3.6.1

### Python version

3.8

### Steps to Reproduce

1. Click `Devices`.
2. Click `Add` and create a new device called "Foo".
3. Click on `Devices` -> `Inventory Items`.
4. Click `Add`.
5. Create an inventory item named "Bar" and assign the parent device "Foo".
6. Click `Create`.
7. Repeat steps 4-6.

### Expected Behavior

The user should be stopped from creating an inventory item under the same parent device with the same name according to the documentation: https://demo.netbox.dev/static/docs/models/dcim/inventoryitem/

### Observed Behavior

Two inventory items with the name "Bar" and the parent device "Foo" exist in the database.

**Comment by abhi1693** [comment]
Thank you for opening a bug report. I was unable to reproduce the reported behavior on NetBox v3.6.1. Please re-confirm the reported behavior on the current stable release and adjust your post above as necessary. Remember to provide detailed steps that someone else can follow using a clean installation of NetBox to reproduce the issue. Remember to include the steps taken to create any initial objects or other data.

**Comment by aaron-iles** [comment]
I'm not sure how else to improve the instructions to make it clearer. I tested this on 3.6.1 several times and can confirm the behavior. 

Here is a screenshot from https://demo.netbox.dev/
![image](https://github.com/netbox-community/netbox/assets/35903275/b598eb84-b1b2-48fb-ad46-00ec17999eb8)

**Comment by abhi1693** [comment]
I checked your work on demo and it has no parent assigned to the items. Whereas your bug report was when the same parent is assigned the code still let's you add them. Please let me know when you have a bug report which is reproducible

**Comment by josh-d-smith** [comment]
Chiming in here since I've encountered the same bug. I think there's confusion here as @aaron-iles is referring to the inventory item `device` field and not the `parent` field.

The below screenshot shows that an inventory item with the same name can be added multiple times to the same `device`. Please note, `parent` is not the issue here.
![netbox_duplicate_names](https://github.com/netbox-community/netbox/assets/122414633/42239d95-fa6c-4ae2-b376-50666c8f3a14)

### Steps to Reproduce (Modified/Clarified)

1. Click `Devices`.
2. Click `Add` and create a new device called `Foo`.
3. Click on `Devices` -> `Inventory Items`.
4. Click `Add`.
5. **_Create an inventory item named `Bar` and assign it to the device `Foo`._**
6. Click `Create`.
7. Repeat steps 4-6.

**Comment by jeremystretch** [comment]
This is happening because PostgreSQL considers NULL values to be unique. We need to add a secondary uniqueness constraint for items which have no parent item assigned.

**Comment by PieterL75** [comment]
Why is there a requirement to have the name of an Inventory item unique ?
We create an item for each CPU, DISK, MEM, ... 
If I have 4 CPU's, I will have 4 times the same name

for swtiches the same.. If I have 4 fans, I will have 4 times the same inventory item

Can this requirement be reconsidered ?

Actually, someone was showing his integration of Norin with netbox : https://netdev-community.slack.com/archives/C01PFCBBQ4T/p1695144834403849?thread_ts=1695135414.451869&cid=C01PFCBBQ4T
and he'll have issues too

**Comment by abhi1693** [comment]
In real world use-case such as yours, CPUs are labelled like CPU 1, CPU 2 etc and the same is for DISK, MEM etc. No component in your device will be unlablled by the manufacturer. Now, whether a user utilizes this information or not is a separate matter IMO.

**Comment by jeremystretch** [comment]
> This will necessitate adding unneeded information to the SFP name just to satisfy the uniqueness constraint.

The only change here is enforcing the same uniqueness requirement for items both with and without a parent item. It would make no sense to treat these differently, hence the bug report.

> Also this is a breaking change in a patch release

This is a bug fix, which are routinely resolved in patch releases per our release policy.

> If the uniqueness constraint is to stay, it would make much more sense to extend it to include the component field as well.

You are welcome to submit a feature request proposing this.

**Comment by kkthxbye-code** [comment]
@jeremystretch - I'll skip the migration instead. I strongly disagree with the reasoning, but I appreciate the response.

**Comment by jeremystretch** [comment]
Maybe a recap would be helpful.

### Prior to merging #13786

The following constraint was in place on InventoryItem:

```python
        constraints = (
            models.UniqueConstraint(
                fields=('device', 'parent', 'name'),
                name='%(app_label)s_%(class)s_unique_device_parent_name'
            ),
        )
```

At first glance, one would assume that this enforces a unique combination of (device, parent, name) for each inventory item, and that is partially correct. For example, within one device, you cannot have two items named "Item1" _if_ they are assigned to the same parent item.

The bug reported here affects items that are not assigned to a parent. This is because PostgreSQL evaluates NULL as not equal to NULL. This result is that uniqueness is enforced for items with a parent and not for items without a parent, which is inconsistent and unpredictable.  **This discrepancy must be resolved** one way or another.

### After merging #13786

The resolution in PR #13786 was to add a second constraint, which ignored null parent assignments when enforcing uniqueness:

```python
        constraints = (
            models.UniqueConstraint(
                fields=('device', 'parent', 'name'),
                name='%(app_label)s_%(class)s_unique_device_parent_name'
            ),
            models.UniqueConstraint(
                fields=('device', 'name'),
                name='%(app_label)s_%(class)s_unique_device_name',
                condition=Q(parent__isnull=True)
            ),
        )
```

This replicates the same approach we've taken elsewhere within the application to ensure consistent uniqueness enforcement. With this change in place, all parentless items assigned to a particular device must have a unique name.

Now, if someone wants to propose removing the uniqueness constraint _entirely_, that may also be considered as an alterntive solution, but it's one which would warrant adequate discussion and consideration as an intentional change versus a correction to intended behavior.

**Comment by jeremystretch** [comment]
I've reverted PR #13741 and re-opened this issue pending discussion on #13837.

**Comment by jeremystretch** [comment]
Given the degree of opposition to this "fix" both here and on #13837, I'm going to flip it around. We'll update the documentation to reflect the current behavior, and treat it as correct until explicitly changed under #13837 or some other future FR.
